# Data wrangling

## Script contents:
### Finding missing values
### Renaming columns
### Creating header rows
### Changing data types
### Resetting indexes

In [1]:
# Import libraries and data files
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\anon\Documents\CareerFoundry\Data Analytics Immersion\Instacart Basket Analysis'

In [3]:
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
df_prods = df = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Drop the 'eval_set column' from orders.csv
df_ords = df_ords.drop(columns = ['eval_set'])

In [6]:
# Find missing values in 'days_since_prior_order' column
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [7]:
# Rename the 'order_dow column' to 'orders_day_of_week'
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [8]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [9]:
# Change 'order_id' column type to string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [10]:
df_ords['order_id'].dtype

dtype('O')

In [11]:
# Import departments csv
df_deps = df = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'departments.csv'), index_col = False)

In [12]:
df_deps.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [13]:
# Transpose departments dataframe from wide to long format and saving to new variable
df_deps_t = df_deps.T

In [14]:
df_deps_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [15]:
# Resetting the index of the departments data
df_deps_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [16]:
# Creating a new header row in the departments data
new_header = df_deps_t.iloc[0]

In [17]:
new_header

0    department
Name: department_id, dtype: object

In [18]:
# Delete existing header row in the departments data
df_deps_t_new = df_deps_t[1:]

In [19]:
df_deps_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [20]:
# Set new header in new dataframe
df_deps_t_new.columns = new_header

In [21]:
df_deps_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [22]:
# Creating a data dictionary
data_dict = df_deps_t_new.to_dict('index')

In [23]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [24]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [25]:
# Creating a subset of the products data set
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [26]:
df_snacks = df_prods[df_prods['department_id']==19]

In [27]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [28]:
df_snacks_2 = df_prods.loc[df_prods['department_id'].isin([19])]

In [29]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# Exercise 4.4

In [30]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## 2) Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [31]:
# Change 'order_number' column type to string
df_ords['order_number'] = df_ords['order_number'].astype('str')

In [32]:
df_ords['order_number'].dtype

dtype('O')

## 3) Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [33]:
# Change the name of the 'order_hour_of_day' column to 'time_order_placed'
df_ords.rename(columns = {'order_hour_of_day' : 'time_order_placed'}, inplace = True)

In [34]:
df_ords.head()

order_id  user_id order_number  orders_day_of_week  time_order_placed  \
0  2539329        1            1                   2                  8   
1  2398795        1            2                   3                  7   
2   473747        1            3                   3                 12   
3  2254736        1            4                   4                  7   
4   431534        1            5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## 4) Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [35]:
# Find the busiest hour for placing orders
result = zip([df_ords['time_order_placed'].value_counts().idxmax()], [df_ords['time_order_placed'].value_counts().max()])

In [36]:
print(tuple(result))

((10, 288418),)


#### The busiest hour for placing orders is the 10th hour of the day (10:00am)

## 5) Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [37]:
# What is the value of 4 in the "department_id" column in df_prods
print(data_dict.get('4'))

{'department': 'produce'}


#### Department 4 is produce

## 6) The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [38]:
# Create a subset containing only the required information about breakfast item sales
df_deps_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [39]:
df_bfast = df_prods.loc[df_prods['department_id']==14]

In [40]:
df_bfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

## 7) Find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [41]:
# Find all observations that include items from the following departments: alcohol, deli, beverages, and meat/seafood
filtered_df = df_deps_t_new[df_deps_t_new['department'].isin(['alcohol', 'deli', 'beverages', 'meat seafood'])]

In [42]:
filtered_df

department_id    department
5                   alcohol
7                 beverages
12             meat seafood
20                     deli

In [43]:
df_dinner_party_items = df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])]

In [44]:
df_dinner_party_items.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

## 8) It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [45]:
df_dinner_party_items.shape

(7650, 5)

#### This new dataframe has 7,650 rows

## 9) Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [46]:
# Extracting all info about user_id = 1
user_id_1 = df_ords[df_ords['user_id'] == 1]

## 10) You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [47]:
# Basic summary stats for user_id = 1
user_id_1.describe()

user_id  orders_day_of_week  time_order_placed  days_since_prior_order
count     11.0           11.000000          11.000000               10.000000
mean       1.0            2.636364          10.090909               19.000000
std        0.0            1.286291           3.477198                9.030811
min        1.0            1.000000           7.000000                0.000000
25%        1.0            1.500000           7.500000               14.250000
50%        1.0            3.000000           8.000000               19.500000
75%        1.0            4.000000          13.000000               26.250000
max        1.0            4.000000          16.000000               30.000000

## 12) Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [48]:
df_ords.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'))

## 13) Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [49]:
df_deps_t_new.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'departments_wrangled.csv'))